saq/queue/postgres.py

```diff
@@ -0,0 +1,447 @@
+
+        await self.notify(job)
+        logger.info("Finished %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+
+    async def dequeue(self, timeout: float = 0) -> Job | None:
+        # Timeout of 0 means timeout immediately
+        result = None
+        conn = await self.pool.getconn(timeout=timeout or None)
+        async with conn.cursor() as cursor:
+            while True:
+                await cursor.execute(
+                    f"""
+                    UPDATE {self.jobs_table} SET status = 'active'
+                    WHERE key = (
+                      SELECT key
+                      FROM {self.jobs_table}
+                      WHERE status = 'queued'
+                        AND queue = %(queue)s
+                        AND %(now)s >= (queued + scheduled)
+                      FOR UPDATE SKIP LOCKED
+                      LIMIT 1
+                    ) AND pg_try_advisory_lock(lock_id)
+                    RETURNING job
+                    """,
+                    {
+                        "queue": self.name,
+                        "now": math.ceil(seconds(now())),
+                    },
+                )
+                result = await cursor.fetchone()
+                if result:
+                    break
+                await asyncio.sleep(0.05)
+        if result:
+            job = self.deserialize(result[0])
+            if job:
+                await conn.commit()
+                self.connections[job.key] = conn
+                return job
+        await conn.rollback()
+        await self.pool.putconn(conn)
+        await asyncio.sleep(0.1)
+        return None
+
+    async def enqueue(self, job_or_func: str | Job, **kwargs: t.Any) -> Job | None:
+        job = self._create_job_for_enqueue(job_or_func, **kwargs)
+
+        await self._before_enqueue(job)
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                f"""
+                INSERT INTO {self.jobs_table} (key, job, queue, status, scheduled)
+                VALUES (%(key)s, %(job)s, %(queue)s, %(status)s, %(scheduled)s)
+                ON CONFLICT (key) DO UPDATE
+                SET job = %(job)s, queue = %(queue)s, status = %(status)s, scheduled = %(scheduled)s
+                WHERE {self.jobs_table}.status NOT IN ('queued', 'aborting')
+                RETURNING job
+                """,
+                {
+                    "key": job.key,
+                    "job": self.serialize(job),
+                    "queue": self.name,
+                    "status": job.status,
+                    "scheduled": job.scheduled,
+                },
+            )
+
+            if not await cursor.fetchone():
+                return None
+
+        logger.info("Enqueuing %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+        return job
+
+    async def delete(
+        self, job: Job, status: Status | None = None, delay: float = 0
+    ) -> None:
+        if delay:
```

Comment:
why do we need a delay for delete?

Reply:
This is tied to the ttl for removing completed (including aborted) tasks. I'll add a janitor task to replace this though.